### PR TITLE
Add CI with GitHub Actions

### PR DESCRIPTION
**Pull request by bksaiki**
This PR adds CI to Ruler. The workflow runs Rust unit tests, synthesizes rules for a few domains, and compares their proving power against previously generated rule sets. In addition, the build process will no longer statically link to Z3, so users need to install `libz3` on their system (users can manually edit the `Cargo.toml` to revert this behavior). The README has been simplified and updated to reflect these changes while the original OOPSLA documentation now in the OOPSLA21.md file.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+on: [push]
+
+env:
+  RUST_BACKTRACE: full
+
+jobs:
+  unit:
+    name: "Unit Tests"
+    runs-on: ubuntu-latest
```
Comment (chandrakananandi) [inline .github/workflows/tests.yml:11]: is this what warfa has?

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+      - name: "Build"
+        run: cargo build --release
+
+      - name: "Unit tests"
+        run: cargo test --release
+
+      - name: "Synthesize rational rules"
+        run: |
+          cargo rational --iters 2 --variables 3 --num-fuzz 50 --do-final-run --outfile rat.json
```
Comment (chandrakananandi) [inline .github/workflows/tests.yml:42]: i think it's worth saving the result jsons in a designated directory.

Comment (chandrakananandi) [reply .github/workflows/tests.yml:42]: I know there is `tests/` but that's the expected ones right? What if we make a `tests/expected` and `tests/actual`? Thoughts?

File: src/bin/real-rat.rs [hunk 3]
```diff
@@ -118,7 +166,7 @@ define_language! {
         "-R" = RSub([Id; 2]),
         "*R" = RMul([Id; 2]),
         "/R" = RDiv([Id; 2]),
-        Var(egg::Symbol),
+        Var(Variable),
```
Comment (chandrakananandi) [inline src/bin/real-rat.rs:169]: why this change?

File: src/derive.rs [hunk 4]
```diff
@@ -9,45 +9,68 @@ pub fn parse<L: SynthLanguage>(filename: &String) -> Vec<Pair<L>> {
...
-            "not_derivable": pairs_to_eqs(&rev_not_derivable),
-        },
-    });
+    let (pairs1, pairs2) = if params.new_eqs {
+        (parse_new_eqs::<L>(&params.in1), parse_new_eqs::<L>(&params.in2))
+    } else {
+        (parse::<L>(&params.in1), parse::<L>(&params.in2))
+    };
+
+    if params.ci {
```
Comment (chandrakananandi) [inline src/derive.rs:40]: yup this seems good!

**Review by chandrakananandi — COMMENTED** [review]
left some questions and comments.

File: .github/workflows/tests.yml (hunk 2 above)
Comment (chandrakananandi) [reply .github/workflows/tests.yml:42]: or maybe we can call it `out/` for the new ones and `tests/` for the pregenerated ones..

File: scripts/generate-tests.sh [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
```
Comment (chandrakananandi) [inline scripts/generate-tests.sh:1]: where is this script run? Just once to generate the "original" results I suppose?

File: .github/workflows/tests.yml (hunk 1 above)
Comment (bksaiki) [reply .github/workflows/tests.yml:11]: This is just CI, not nightlies. We would need a considerably more complicated system for that.

File: .github/workflows/tests.yml (hunk 2 above)
Comment (bksaiki) [reply .github/workflows/tests.yml:42]: Makes sense to me. I'll probably go with `out/` and `tests/`.

File: scripts/generate-tests.sh (hunk 5 above)
Comment (bksaiki) [reply scripts/generate-tests.sh:1]: Yep. It should only be rerun, if we want to change our expected (baseline) rulesets.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (chandrakananandi) [reply .github/workflows/tests.yml:11]: oh right sorry I was still thinking of nightlies in my head ever since we discussed it in the egraph channel.

File: src/bin/real-rat.rs (hunk 3 above)
Comment (bksaiki) [reply src/bin/real-rat.rs:169]: Parsing issue. It was actually causing derive to crash.

Comment (chandrakananandi) [reply src/bin/real-rat.rs:169]: weird, ok!

**Review by chandrakananandi — APPROVED** [review]
Thanks Brett!!
